### ISSUE TITLE
[Ruby] Fix code folding for heredoc

### ISSUE THREAD

**Issue by zufuliu** [ScintillaOrg/lexilla#135]
current code fails when here are space after opening delimiter due to opening delimiter is not skipped after increase fold level:
```ruby
a = <<XXX # :nodoc:
heredoc
XXX
```

```diff
diff --git a/lexers/LexRuby.cxx b/lexers/LexRuby.cxx
index 2733dcb3..f61c3450 100644
--- a/lexers/LexRuby.cxx
+++ b/lexers/LexRuby.cxx
@@ -1835,6 +1834,7 @@ void FoldRbDoc(Sci_PositionU startPos, Sci_Position length, int initStyle,
     };
     MethodDefinition method_definition = MethodDefinition::None;
     int argument_paren_count = 0;
+    bool heredocOpen = false;
 
     for (Sci_PositionU i = startPos; i < endPos; i++) {
         const char ch = chNext;
@@ -1895,10 +1895,15 @@ void FoldRbDoc(Sci_PositionU startPos, Sci_Position length, int initStyle,
                 levelCurrent++;
             }
         } else if (style == SCE_RB_HERE_DELIM) {
-            if (styler.SafeGetCharAt(i-2) == '<' && styler.SafeGetCharAt(i-1) == '<') {
+            if (stylePrev == SCE_RB_OPERATOR && chPrev == '<' && styler.SafeGetCharAt(i - 2) == '<') {
                 levelCurrent++;
-            } else if (styleNext == SCE_RB_DEFAULT) {
-                levelCurrent--;
+                heredocOpen = true;
+            } else if (styleNext != SCE_RB_HERE_DELIM) {
+                if (heredocOpen) {
+                    heredocOpen = false;
+                } else {
+                    levelCurrent--;
+                }
             }
         } else if (style == SCE_RB_STRING_QW || style == SCE_RB_STRING_W) {
             if (stylePrev != style) {
```

**Comment by nyamatongwe** [comment]
This appears to break folding of `test/examples/ruby/Issue66.rb` by continuing first fold past end of heredoc.

~~~ruby
<<T
X
T

<<-T
X
T
~~~

**Comment by zufuliu** [comment]
[ruby-0216-135.zip](https://github.com/ScintillaOrg/lexilla/files/10754714/ruby-0216-135.zip)

Updated patch, also fix following code (currently unsupported, called `stacked here-docs` in Perl lexer):
https://docs.ruby-lang.org/en/master/syntax/literals_rdoc.html#label-Here+Document+Literals
> You may open multiple heredocs on the same line, but this can be difficult to read:

```ruby
puts(<<-ONE, <<-TWO)
content for heredoc one
ONE
content for heredoc two
TWO
```

**Comment by nyamatongwe** [comment]
Included in 5.2.3.
